### PR TITLE
[receiver/httpcheck] add metrics check whether substring contains in response body

### PR DESCRIPTION
**Pull request by sakulali**
**Description:**
Part of https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/24913, add metrics check if substring contains in http check response body or not. When set `body` option and emit `httpcheck.body`,  testing substring contains or not in the response of HTTP Check. 

**Link to tracking Issue:**
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/24913

**Testing:**
make generate
make chlog-validate
go test for httpcheckreceiver

**Documentation:**
receiver/httpcheckreceiver/README.md

### CODE REVIEW

File: receiver/httpcheckreceiver/scraper.go
Comment (atoulme) [inline receiver/httpcheckreceiver/scraper.go:98]: So if the body is empty we don't report that there's no match?

File: receiver/httpcheckreceiver/scraper.go
Comment (atoulme) [inline receiver/httpcheckreceiver/scraper.go:103]: I'm not sure why you'd need both a metric for content matched and a metric for content not matched, given that you're able to capture that with the 0 and 1 values. Can you simplify to just one metric to match content.

File: receiver/httpcheckreceiver/metadata.yaml
Comment (atoulme) [inline receiver/httpcheckreceiver/metadata.yaml:60]: match has a connotation of regular expression, I'd use the term "substring" instead of match_content.

**Comment by lpegoraro** [comment]
Hi @sakulali, if you need help with this, let me know, I can work with you, since this is aligned with what I need for the receiver.

**Comment by sakulali** [comment]
> Hi @sakulali, if you need help with this, let me know, I can work with you, since this is aligned with what I need for the receiver.

Sorry for delay reply @lpegoraro. Could you please help review the code? I'm unsure if my implementation aligns with your issue, and I would also like to collaborate in completing this valuable feature you've proposed.

**Comment by sakulali** [comment]
Hello @atoulme, cloud you mind help reviews again? I want to work with @lpegoraro  to implement the feature, thanks!

File: .chloggen/httpcheckreceiver-match-content.yaml
Comment (atoulme) [inline .chloggen/httpcheckreceiver-match-content.yaml:10]:
```suggestion
note: Add metrics testing if the HTTP response matches a specific substring.
```

File: receiver/httpcheckreceiver/README.md
Comment (atoulme) [inline receiver/httpcheckreceiver/README.md:40]:
```suggestion
- `substring` (optional, default: ""): If set, the receiver will emit metrics based on whether the response body matches this string.
```

File: receiver/httpcheckreceiver/metadata.yaml
Comment (atoulme) [inline receiver/httpcheckreceiver/metadata.yaml:61]:
```suggestion
    description: Records whether the HTTP response body matches the `substring` configuration item.
```

File: receiver/httpcheckreceiver/metadata.yaml
Comment (atoulme) [inline receiver/httpcheckreceiver/metadata.yaml:32]: Is this really important? This value will be constant based on configuration.

File: receiver/httpcheckreceiver/scraper.go
Comment (atoulme) [inline receiver/httpcheckreceiver/scraper.go:98]: I note here that `string(body)` does not care for encoding. Not sure if this is a problem.

**Comment by atoulme** [comment]
Some English and a couple touches, but it's not far.

File: receiver/httpcheckreceiver/metadata.yaml
Comment (sakulali) [reply receiver/httpcheckreceiver/metadata.yaml]: yes, you are right, this value will be constant and i need to remove this attribute.

File: receiver/httpcheckreceiver/scraper.go
Comment (sakulali) [reply receiver/httpcheckreceiver/scraper.go:98]: Very detailed observation! At that time, i was only considering plain text responses for HTTP, without considering encoding formats like base64 or others. If we need to consider encoding formats, we may have to take into account multiple encoding methods?

File: .chloggen/httpcheckreceiver-match-content.yaml
Comment (sakulali) [reply .chloggen/httpcheckreceiver-match-content.yaml]: Done

File: receiver/httpcheckreceiver/README.md
Comment (sakulali) [reply receiver/httpcheckreceiver/README.md]: Done

File: receiver/httpcheckreceiver/metadata.yaml
Comment (sakulali) [reply receiver/httpcheckreceiver/metadata.yaml]: Done

**Comment by sakulali** [comment]
> Some English and a couple touches, but it's not far.

Thanks very much for taking the time to reviews @atoulme!

File: receiver/httpcheckreceiver/scraper.go
Comment (atoulme) [reply receiver/httpcheckreceiver/scraper.go:98]: yeah I am all for you to make this incremental improvement, and we can always go back and learn more about what to expect. I am more worried of double byte encoded strings such as UTF-16 encoded strings.

**Review by atoulme — APPROVED** [review]

File: receiver/httpcheckreceiver/scraper.go
Comment (lpegoraro) [reply receiver/httpcheckreceiver/scraper.go:98]: we could maybe check the encoding on the header?

Comment (lpegoraro) [reply receiver/httpcheckreceiver/scraper.go:98]: and based on the header add the decode method? like a switch

Comment (atoulme) [reply receiver/httpcheckreceiver/scraper.go:98]: here we go, that's how it should be done. That's probably something to keep in mind for a subsequent PR.

**Comment by lpegoraro** [comment]
I also think we are very close!
I only have a suggestion about the encoding of the string of the check, if we either get the encoding from the configuration of the target, or do we fetch it from the Header, which I believe is safer, what do you guys think?

**Comment by sakulali** [comment]
> if we either get the encoding from the configuration of the target, or do we fetch it from the Header

Great suggestion! I think it's a good idea to create a separate issue to follow up. I'm not very familiar with encoding best practices either, i think we can retrieve the encoding format from a custom field in the HTTP header, such as `X-Encoding: my-custom-encoding`, or through URL parameters. We can then use the configuration file to select the appropriate decoding method for processing?

**Review by lpegoraro — APPROVED** [review]

File: receiver/httpcheckreceiver/README.md
Comment (codeboten) [inline receiver/httpcheckreceiver/README.md:40]: As per the description of this option, `body` seems more descriptive

```suggestion
- `body` (optional, default: ""): If set, the receiver will emit metrics based on whether the response body matches this string.
```

File: receiver/httpcheckreceiver/config.go
Comment (codeboten) [inline receiver/httpcheckreceiver/config.go:34]:
```suggestion
	Body                          string `mapstructure:"body"`
```

File: receiver/httpcheckreceiver/metadata.yaml
Comment (codeboten) [inline receiver/httpcheckreceiver/metadata.yaml:57]:
```suggestion
  httpcheck.body:
```

File: receiver/httpcheckreceiver/metadata.yaml
Comment (codeboten) [inline receiver/httpcheckreceiver/metadata.yaml:58]:
```suggestion
    description: Reports 1 if the HTTP response body matches the `body` configuration item, 0 otherwise.
```

**Review by codeboten — COMMENTED** [review]
Thanks for the submission @sakulali. I would suggest renaming substring to body. I wonder if a new metric is the right semantic here, or if instead it should be an attribute of an existing metric.

@astencel-sumo any thoughts?

File: receiver/httpcheckreceiver/README.md
Comment (astencel-sumo) [reply receiver/httpcheckreceiver/README.md:40]: I agree `body` is much better name for both the configuration property and the metric name.

I would also expect the description to specify whether it's a regex match or exact match.

File: receiver/httpcheckreceiver/metadata.yaml
Comment (astencel-sumo) [reply receiver/httpcheckreceiver/metadata.yaml:57]: > I wonder if a new metric is the right semantic here, or if instead it should be an attribute of an existing metric.

@codeboten this is a very good question. As a user, my assumption would be that if I specified a body match string in a target's configuration, then I only get `httpcheck.status` metrics with the value `1` if the body matched, and I'd expect getting `httpcheck.status` with value `0` and possibly an additional attribute e.g. `reason=body` or similar, explaining why the check failed.

I'm not sure how this fits the current design of the receiver, in that it reports a separate data point for each `http.status_class` (1xx, 2xx, etc.). By the way, I'm not sure what advantages this has over reporting just a single `httpcheck.status` data point for each target and request.

File: receiver/httpcheckreceiver/metadata.yaml
Comment (astencel-sumo) [inline receiver/httpcheckreceiver/metadata.yaml:65]: Shouldn't we add the body match string (the value of `substring`) as an attribute of this metric?

Let's imagine the following scenario: I have a web page that reports status of several things, and I want to use the HTTP Check receiver to check the status of those things. For example, it's a web page that has the following content:

```html
Component 1: OK
Component 2: Failed
Component 3: OK
```

My HTTP Check receiver configuration would then be:

```yaml
receivers:
  httpcheck:
    targets:
    - endpoint: http://my-web.page
      substring: "Component 1: OK"
    - endpoint: http://my-web.page
      substring: "Component 2: OK"
    - endpoint: http://my-web.page
      substring: "Component 3: OK"
```

With the proposed design of the `httpcheck.substring` metric, I would get three data points with the same name and attributes and couldn't differentiate between them:

```
httpcheck.substring{http.url="http://my-web.page"} 1
httpcheck.substring{http.url="http://my-web.page"} 0
httpcheck.substring{http.url="http://my-web.page"} 1
```

Another way to fix this might be to add a `name` property to the `target` definition, that could be used as an attribute.

Not sure if this example isn't stretching what the HTTP Check receiver is design to do though. @codeboten any thoughts?

**Comment by sakulali** [comment]
Thanks very much for your thorough reviews @codeboten! @astencel-sumo!  I have already renamed `substring` to `body`,  we need to confirm again whether body exact match string as an attribute of the `httpcheck.body` metric is required as @astencel-sumo mentioned?

File: receiver/httpcheckreceiver/scraper.go
Comment (astencel-sumo) [inline receiver/httpcheckreceiver/scraper.go:93]: Would be good to mention which target this error applies to, wouldn't it?

Comment (sakulali) [reply receiver/httpcheckreceiver/scraper.go]: Thanks @astencel-sumo, that's a great suggestion.

**Comment by lpegoraro** [comment]
hi there folks, what is the hold up? can we do something about it?

**Comment by sakulali** [comment]
> hi there folks, what is the hold up? can we do something about it?

Thanks for reminder. Could you mind help to reviews again @codeboten

File: receiver/httpcheckreceiver/metadata.yaml
Comment (atoulme) [reply receiver/httpcheckreceiver/metadata.yaml:65]: I agree with your point and I propose an optional `name` attribute. The default value would be an empty string, in which case the name attribute would be omitted.

Comment (sakulali) [reply receiver/httpcheckreceiver/metadata.yaml:65]: Thanks for your suggestion @atoulme! I have a doubt about how to implement the optional functionality. Do you mean add `name` attribute in targetConfig(Default is empty) and add `name` metric attribute?
``` Go
// receiver/httpcheckreceiver/config.go
type targetConfig struct {
	confighttp.HTTPClientSettings `mapstructure:",squash"`
	Method                        string `mapstructure:"method"`
	Body                          string `mapstructure:"body"`
	Name 	                      string `mapstructure:"name"`
}
```
``` Yaml
# receiver/httpcheckreceiver/metadata.yaml
attributes:
  http.url:
    description: Full HTTP request URL.
    type: string
  name:
    description: Body exact matches string.
    type: string
```

Comment (atoulme) [reply receiver/httpcheckreceiver/metadata.yaml:65]: yes

Comment (sakulali) [reply receiver/httpcheckreceiver/metadata.yaml]: Done. Would it be possible for help to review it again when you have some free time @atoulme and @astencel-sumo? Thanks!

File: receiver/httpcheckreceiver/metadata.yaml
Comment (atoulme) [inline receiver/httpcheckreceiver/metadata.yaml:31]:
```suggestion
    description: Optional name of this check.
```

File: receiver/httpcheckreceiver/README.md
Comment (atoulme) [inline receiver/httpcheckreceiver/README.md:41]:
```suggestion
- `name` (optional, default: ""): The name of this receiver.
```

File: receiver/httpcheckreceiver/metadata.yaml
Comment (sakulali) [reply receiver/httpcheckreceiver/metadata.yaml]: Done

File: receiver/httpcheckreceiver/README.md
Comment (sakulali) [reply receiver/httpcheckreceiver/README.md]: Done

File: receiver/httpcheckreceiver/metadata.yaml
Comment (astencel-sumo) [reply receiver/httpcheckreceiver/metadata.yaml]: Marking this conversation as Resolved, as it is resolved from my perspective.

File: receiver/httpcheckreceiver/metadata.yaml
Comment (astencel-sumo) [reply receiver/httpcheckreceiver/metadata.yaml]: Thanks @sakulali for all the changes. I still think the HTTP Check receiver should only emit a single metric, and a single data point for every target and check (HTTP request). This pull request adds another metric `httpcheck.body` to the already existing metric `httpcheck.status` that has multiple data points (one for every HTTP response code class - 1xx, 2xx, etc.).

This is not the direction I want to see for this receiver (see https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/22994), therefore I'm not going to approve this pull request. I'm also not going to stop this pull request from being merged if others approve it.

**Review by codeboten — APPROVED** [review]

File: receiver/httpcheckreceiver/metadata.yaml
Comment (sakulali) [reply receiver/httpcheckreceiver/metadata.yaml]: Thanks for your detailed response @astencel-sumo! After reviewing the issue you mentioned, maybe i understand your concerns. I have a few thoughts and a small question:

1. Regarding the httpcheck.status metric, if there is only one data point, it seems that the value is always 1. In that case, how can we track the count of response codes in the 1xx, 2xx, 3xx, etc range?

2. If we don't add a separate metric `httpcheck.body`, instead use the body match as a condition for `httpcheck.status`, then the code will need further adjustments.

3. If we decide to keep only the httpcheck.status metric with a default single data point, I believe we would need a feature gate to enable or disable this functionality.

Since I'm not very familiar with the design of the HTTP Checker and not quite clear about the specific user scenarios, i would like to reach a consensus on a reasonable solution.

**Comment by lpegoraro** [comment]
Hi there! Great work on getting 2 approvals, can we merge this for the next release?

**Comment by codeboten** [comment]
Sorry for the delay @lpegoraro, I would like to resolve https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/27015#discussion_r1420993982 before merging this

File: receiver/httpcheckreceiver/metadata.yaml
Comment (lpegoraro) [reply receiver/httpcheckreceiver/metadata.yaml]: I also believe that this can be an attribute of the result.
Putting it an example I think it's better for understandind.
Like if you are looking for something like a matching version on the target
the httpcheck.status = 200, but the httpcheck.body = false 
But when the version is updated, then the httpcheck.body = true.

**Comment by lpegoraro** [comment]
Count me in for anything @codeboten

File: receiver/httpcheckreceiver/README.md
Comment (tusharbhasme) [inline receiver/httpcheckreceiver/README.md:40]: This could carry a different meaning when method is `POST` where the `body` content will be sent as part of the request.

**Comment by AwsmAshraf** [comment]
Any one working on this? I actually have a requirement for this. Can we do it in a way in which we can specify the status code to be expected? I actually agree with @andrzej-stencel on this. A single data point would be great, with any additional info in the form of attributes. Also, modifying the name of the metric to something other than `httpcheck.status` would make identifying the result of multiple http checks a lot easier. What I'm trying to say is different http endpoint checks would be populated as different metrics with the name of the metric specified for that endpoint when configuring this receiver.

**Pull request closed without merging**
